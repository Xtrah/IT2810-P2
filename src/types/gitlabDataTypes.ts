/* eslint-disable camelcase */ // Ignore because of API is not set to camelcase
export interface Commit {
  author_name: string;
  created_at: string;
}

export interface Issue {
  closed_at: string | null;
  created_at: string;
}
