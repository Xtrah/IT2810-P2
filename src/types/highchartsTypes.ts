interface SeriesContent {
  name: string;
  data: number[];
}

export type Series = SeriesContent[];
