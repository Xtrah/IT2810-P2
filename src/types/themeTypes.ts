/* eslint-disable no-unused-vars */
/* eslint-disable no-shadow */
/* NB! Disabled because of eslint bug NB! */
/* https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md#i-am-using-a-rule-from-eslint-core-and-it-doesnt-work-correctly-with-typescript-code */
enum ThemeTypes {
  DARK = "dark",
  LIGHT = "light",
}

export default ThemeTypes;
