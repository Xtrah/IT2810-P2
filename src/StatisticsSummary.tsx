import styled from "styled-components";
import { PersonCircle } from "@styled-icons/bootstrap/PersonCircle";
import { CheckCircle } from "@styled-icons/bootstrap/CheckCircle";
import { ExclamationCircle } from "@styled-icons/bootstrap/ExclamationCircle";
import { ChangeEvent } from "react";
import Section from "./components/Section";
import { Commit, Issue } from "./types/gitlabDataTypes";
import getUniqueCommitContributors from "./utils/getUniqueCommitContributors";
import useSessionStorage from "./utils/useSessionStorage";

const DataParagraph = styled.p`
  display: inline;
  font-size: 24px;
`;

const SummaryTitle = styled.h1`
  font-size: 14px;
  color: grey;
  margin: 0;
  margin-bottom: 1em;
`;

const SummaryList = styled.ul`
  list-style: none;
  padding: 0;
  margin: 0;
`;

const CheckIcon = styled(CheckCircle)`
  vertical-align: sub;
`;

const UserIcon = styled(PersonCircle)`
  vertical-align: sub;
`;

const ExclamationIcon = styled(ExclamationCircle)`
  vertical-align: sub;
`;

const ColoredSpan = styled.span`
  color: orange;
  margin: 0.25em;
`;

const Text = styled.p`
  margin-bottom: 4px;
`;

interface Props {
  issuesData: Issue[];
  commitsData: Commit[];
  // eslint-disable-next-line no-unused-vars
  onChange: (dateOption: number) => void;
}

function StatisticsSummary({ issuesData, commitsData, onChange }: Props) {
  const initialDaysToInclude = "99999";
  const [daysToIncludeData, setDaysToIncludeData] = useSessionStorage(
    "daysToInclude",
    initialDaysToInclude
  );

  const handleChange = (e: ChangeEvent<HTMLSelectElement>) => {
    const newVal = e.target.value;
    onChange(parseInt(newVal, 10));
    setDaysToIncludeData(newVal);
  };

  return (
    <Section>
      <SummaryTitle>Current status of repository</SummaryTitle>
      <SummaryList>
        <li>
          <ExclamationIcon size={24} fill="orange" />
          <DataParagraph>
            <ColoredSpan>{issuesData.length}</ColoredSpan>Total issues
          </DataParagraph>
        </li>
        <li>
          <CheckIcon size={24} fill="orange" />
          <DataParagraph>
            <ColoredSpan>{commitsData.length}</ColoredSpan>Total commits
          </DataParagraph>
        </li>
        <li>
          <UserIcon size={24} fill="orange" />
          <DataParagraph>
            <ColoredSpan>
              {getUniqueCommitContributors(commitsData).length}
            </ColoredSpan>
            Total contributors
          </DataParagraph>
        </li>
      </SummaryList>

      <Text>Data from </Text>
      <select value={daysToIncludeData} onChange={handleChange}>
        <option value="99999">all time</option>
        <option value="30">last 30 days</option>
        <option value="14">last 14 days</option>
        <option value="7">last 7 days</option>
      </select>
    </Section>
  );
}

export default StatisticsSummary;
