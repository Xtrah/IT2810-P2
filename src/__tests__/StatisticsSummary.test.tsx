import renderer from "react-test-renderer";
import StatisticsSummary from "../StatisticsSummary";
import { Commit, Issue } from "../types/gitlabDataTypes";

// https://blog.openreplay.com/react-snapshot-testing-with-jest-an-introduction-with-examples

it("should render summary with 0 total commits, contributors and issues when empty lists of data", () => {
  const issuesData: Issue[] = [];
  const commitsData: Commit[] = [];

  const elem = renderer
    .create(
      <StatisticsSummary issuesData={issuesData} commitsData={commitsData} />
    )
    .toJSON();

  expect(elem).toMatchSnapshot();
});

it("should render summary with correct count when lists of data", () => {
  const issuesData: Issue[] = [
    { closed_at: "2021-09-21T12:33:09.076+02:00" },
    { closed_at: null },
    { closed_at: "2021-09-21T12:33:09.076+02:00" },
  ];
  const commitsData: Commit[] = [
    { author_name: "user1" },
    { author_name: "user1" },
    { author_name: "user2" },
    { author_name: "user3" },
  ];

  const elem = renderer
    .create(
      <StatisticsSummary issuesData={issuesData} commitsData={commitsData} />
    )
    .toJSON();

  expect(elem).toMatchSnapshot();
});
