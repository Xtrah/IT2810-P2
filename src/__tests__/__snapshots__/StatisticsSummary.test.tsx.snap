// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<StatisticsSummary /> should render summary with 0 total commits, contributors and issues when empty lists of data 1`] = `
<section
  className="sc-bdfBQB goIpOt"
>
  <h1
    className="sc-dlfnuX ivCGQX"
  >
    Current status of repository
  </h1>
  <ul
    className="sc-hKgJUU ckoNhy"
  >
    <li>
      <svg
        aria-hidden="true"
        className="StyledIconBase-ea9ulj-0 iuNRof sc-gKseQn iCzHnw"
        fill="currentColor"
        focusable="false"
        height={34}
        viewBox="0 0 16 16"
        width={34}
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M8 15A7 7 0 1 1 8 1a7 7 0 0 1 0 14zm0 1A8 8 0 1 0 8 0a8 8 0 0 0 0 16z"
        />
        <path
          d="M7.002 11a1 1 0 1 1 2 0 1 1 0 0 1-2 0zM7.1 4.995a.905.905 0 1 1 1.8 0l-.35 3.507a.552.552 0 0 1-1.1 0L7.1 4.995z"
        />
      </svg>
      <p
        className="sc-gsTEea hdYJyo"
      >
        <span
          className="sc-iBPTik hakRGd"
        >
          0
        </span>
        Total issues
      </p>
    </li>
    <li>
      <svg
        aria-hidden="true"
        className="StyledIconBase-ea9ulj-0 iuNRof sc-eCstlR LucOW"
        fill="currentColor"
        focusable="false"
        height={34}
        viewBox="0 0 16 16"
        width={34}
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M8 15A7 7 0 1 1 8 1a7 7 0 0 1 0 14zm0 1A8 8 0 1 0 8 0a8 8 0 0 0 0 16z"
        />
        <path
          d="M10.97 4.97a.235.235 0 0 0-.02.022L7.477 9.417 5.384 7.323a.75.75 0 0 0-1.06 1.06L6.97 11.03a.75.75 0 0 0 1.079-.02l3.992-4.99a.75.75 0 0 0-1.071-1.05z"
        />
      </svg>
      <p
        className="sc-gsTEea hdYJyo"
      >
        <span
          className="sc-iBPTik hakRGd"
        >
          0
        </span>
        Total commits
      </p>
    </li>
    <li>
      <svg
        aria-hidden="true"
        className="StyledIconBase-ea9ulj-0 iuNRof sc-jSgvzq jXfoLi"
        fill="currentColor"
        focusable="false"
        height={34}
        viewBox="0 0 16 16"
        width={34}
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M11 6a3 3 0 1 1-6 0 3 3 0 0 1 6 0z"
        />
        <path
          d="M0 8a8 8 0 1 1 16 0A8 8 0 0 1 0 8zm8-7a7 7 0 0 0-5.468 11.37C3.242 11.226 4.805 10 8 10s4.757 1.225 5.468 2.37A7 7 0 0 0 8 1z"
          fillRule="evenodd"
        />
      </svg>
      <p
        className="sc-gsTEea hdYJyo"
      >
        <span
          className="sc-iBPTik hakRGd"
        >
          0
        </span>
        Total contributors
      </p>
    </li>
  </ul>
  <p
    className="sc-fubCzh fqkfqU"
  >
    Data from 
  </p>
  <select
    onChange={[Function]}
    value="99999"
  >
    <option
      value="99999"
    >
      all time
    </option>
    <option
      value="30"
    >
      last 30 days
    </option>
    <option
      value="14"
    >
      last 14 days
    </option>
    <option
      value="7"
    >
      last 7 days
    </option>
  </select>
</section>
`;

exports[`<StatisticsSummary /> should render summary with correct count when lists of data 1`] = `
<section
  className="sc-bdfBQB goIpOt"
>
  <h1
    className="sc-dlfnuX ivCGQX"
  >
    Current status of repository
  </h1>
  <ul
    className="sc-hKgJUU ckoNhy"
  >
    <li>
      <svg
        aria-hidden="true"
        className="StyledIconBase-ea9ulj-0 iuNRof sc-gKseQn iCzHnw"
        fill="currentColor"
        focusable="false"
        height={34}
        viewBox="0 0 16 16"
        width={34}
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M8 15A7 7 0 1 1 8 1a7 7 0 0 1 0 14zm0 1A8 8 0 1 0 8 0a8 8 0 0 0 0 16z"
        />
        <path
          d="M7.002 11a1 1 0 1 1 2 0 1 1 0 0 1-2 0zM7.1 4.995a.905.905 0 1 1 1.8 0l-.35 3.507a.552.552 0 0 1-1.1 0L7.1 4.995z"
        />
      </svg>
      <p
        className="sc-gsTEea hdYJyo"
      >
        <span
          className="sc-iBPTik hakRGd"
        >
          3
        </span>
        Total issues
      </p>
    </li>
    <li>
      <svg
        aria-hidden="true"
        className="StyledIconBase-ea9ulj-0 iuNRof sc-eCstlR LucOW"
        fill="currentColor"
        focusable="false"
        height={34}
        viewBox="0 0 16 16"
        width={34}
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M8 15A7 7 0 1 1 8 1a7 7 0 0 1 0 14zm0 1A8 8 0 1 0 8 0a8 8 0 0 0 0 16z"
        />
        <path
          d="M10.97 4.97a.235.235 0 0 0-.02.022L7.477 9.417 5.384 7.323a.75.75 0 0 0-1.06 1.06L6.97 11.03a.75.75 0 0 0 1.079-.02l3.992-4.99a.75.75 0 0 0-1.071-1.05z"
        />
      </svg>
      <p
        className="sc-gsTEea hdYJyo"
      >
        <span
          className="sc-iBPTik hakRGd"
        >
          4
        </span>
        Total commits
      </p>
    </li>
    <li>
      <svg
        aria-hidden="true"
        className="StyledIconBase-ea9ulj-0 iuNRof sc-jSgvzq jXfoLi"
        fill="currentColor"
        focusable="false"
        height={34}
        viewBox="0 0 16 16"
        width={34}
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M11 6a3 3 0 1 1-6 0 3 3 0 0 1 6 0z"
        />
        <path
          d="M0 8a8 8 0 1 1 16 0A8 8 0 0 1 0 8zm8-7a7 7 0 0 0-5.468 11.37C3.242 11.226 4.805 10 8 10s4.757 1.225 5.468 2.37A7 7 0 0 0 8 1z"
          fillRule="evenodd"
        />
      </svg>
      <p
        className="sc-gsTEea hdYJyo"
      >
        <span
          className="sc-iBPTik hakRGd"
        >
          3
        </span>
        Total contributors
      </p>
    </li>
  </ul>
  <p
    className="sc-fubCzh fqkfqU"
  >
    Data from 
  </p>
  <select
    onChange={[Function]}
    value="99999"
  >
    <option
      value="99999"
    >
      all time
    </option>
    <option
      value="30"
    >
      last 30 days
    </option>
    <option
      value="14"
    >
      last 14 days
    </option>
    <option
      value="7"
    >
      last 7 days
    </option>
  </select>
</section>
`;
